### PR TITLE
feat(bindnode): support listpairs struct representation

### PR DESCRIPTION
**Pull request by rvagg**
I want order-preserving structs without a pre-defined strict order (tuple), so this implements listpairs which is in the schema spec but has sparse support through go-ipld-prime.

```ipldsch
type Blop struct {
  foo optional Int
  bar optional String
  baz optional String
} representation listpairs
```

```go
type Blop struct {
  Foo *int
  Bar  *string
  Baz *string
}
```

```json
[ "foo", 100, "baz", "bing" ]
```

_(edit: actually should be this:)_

```json
[ [ "foo", 100 ], [ "baz", "bing" ] ]
```

But also this should decode to the same struct (see `out-of-order` test):

```json
[ "baz", "bing", "foo", 100 ]
```

_(edit: actually should be this:)_

```json
[ [ "baz", "bing" ], [ "foo", 100 ] ]
```

### CODE REVIEW

File: node/bindnode/repr.go [hunk 1]
```diff
@@ -974,14 +1055,24 @@ func (w *_listStructAssemblerRepr) AssembleValue() datamodel.NodeAssembler {
 		}
 		entryAsm = assemblerRepr(entryAsm)
 		return entryAsm
+	case schema.StructRepresentation_ListPairs:
+		idx := w.nextIndex
+		w.nextIndex++
+		if idx%2 == 0 {
+			asm := (*_structAssembler)(w).AssembleKey()
+			// ???
```
[the remark this answers is not in the document]
Comment (rvagg) [reply node/bindnode/repr.go:1063]: oh, embarrassing ... that's there cause I don't actually know why I would want to put the next line there; I intended to figure this out

**Comment by rvagg** [comment]
marked as draft cause I just realised after writing out the example above that listpairs is actually `[[k1,v1],[k2,v2]]`, so a list of k/v tuples; so I need to update the repr to look like that!

**Comment by rvagg** [comment]
OK, fixed, this now handles `[[k1,v1],[k2,v2]...]`. It's required a bit more code because of the need for new intermediate assemblers for the inner lists that are strictly length of two, so we have `_listpairsFieldAssemblerRepr` and `_listpairsFieldListAssemblerRepr` to set that up properly.

I also extended the test a bit more so we're not just setting strings as values but also adding a recursive kind on one of the fields, so we end up doing a `[["foo","0"],["bar","1"],["baz","2"],["qux",["3","4"]]]` where the `qux` field is a list of strings.

**Comment by rvagg** [comment]
Made the parser handle "listpairs" for structs in the DSL, including enabling the test fixture that's been waiting to be implemented: https://github.com/ipld/ipld/pull/281

File: node/bindnode/repr.go [hunk 2]
```diff
@@ -992,6 +1076,116 @@ func (w *_listStructAssemblerRepr) ValuePrototype(idx int64) datamodel.NodeProto
...
+func (w _listpairsFieldAssemblerRepr) AssignLink(datamodel.Link) error {
+	return datamodel.ErrWrongKind{
+		TypeName:        w.parent.schemaType.Name(),
+		MethodName:      "AssignLink",
+		AppropriateKind: datamodel.KindSet_JustList,
+		ActualKind:      datamodel.Kind_Map,
+	}
+}
+
+// TODO: support this if it's a list?
```
Comment (rvagg) [inline node/bindnode/repr.go:1151]: turns out I need to do this, so another blocker to merge

Comment (rvagg) [reply node/bindnode/repr.go]: done

**Review by willscott — APPROVED** [review]
🚀

**Comment by rvagg** [comment]
good thing I held off until I had a chance to properly try and use this, it borked on complex nested values—trying to do a listpairs within a listpairs errored out; very tricky to get the "level" right with these builders, representation builders use the schema builders but always need to force them to drop back down to representation level when returning sub-builders to the user

hopefully fixed now, still trying to put it to use

**Pull request merged**